### DISCUSSION TITLE
Reusing Yii3 packages

### DISCUSSION THREAD

**Discussion by samdark** [yiisoft/yii2#19875] [Yii 22]
One of the ideas for Yii 2.2 was reusing Yii3 packages. That's a nice idea if we'll gradually move from Yii 2.0 because:

1. There will be singnficantly less maintenance for legacy code.
2. Packages such as [yiisoft/db](https://github.com/yiisoft/db) are way better in terms of testing and bugfixes than what's in Yii2 thanks to @terabytesoftw and @darkdef.
3. We'll have compatibility layer so people will be able to use what we've worked on for a long time with way easier migration path.

The hard decision here is to align PHP versions so that will require Yii 2.2 to be PHP 8.0+ at least.

**Comment by samdark** [comment]
What could be reused right away:

- https://github.com/yiisoft/cache — to replace caching.
- https://github.com/yiisoft/db — to replace database-related functionality (could be tricky and raises questions about AR which isn't yet ready).
- https://github.com/yiisoft/rbac — to replace RBAC.
- https://github.com/yiisoft/arrays — to replace array helper (partially).
- https://github.com/yiisoft/log — to replace logger.
- https://github.com/yiisoft/strings — to replace string helper (partially).
- https://github.com/yiisoft/mutex — to replace mutex.
- https://github.com/yiisoft/files — to replace file helper (partially).
- https://github.com/yiisoft/json — to replace JSON helper.
- https://github.com/yiisoft/network-utilities — to replace implementations of network-related stuff.
- https://github.com/yiisoft/http — to replace status codes.
- https://github.com/yiisoft/mailer — to replace mailer abstraction and drivers.
- https://github.com/yiisoft/translator — to replace underlying implementation of `Yii::t()`.
- https://github.com/yiisoft/var-dumper — to replace the corresponding helper.

**Reply by simialbi** [reply]
How do you imagine the result? I think with this approach it results in something like yii 3 base app, just without app, doesn't it?

**Reply by tomaszkane** [reply]
When I compare yii3 demo with yii2 app basic it's totally different, so IMHO results is not something like yii 3 base app, just without app.

**Reply by samdark** [reply]
Same Yii2 with some internal parts replaced.

**Reply by fusivel** [reply]
Just a silly idea:
Wouldn't it be better to create a "rosetta" kind of extension (like apple does for changing processors) that would bridge and take care of all the traits, instead of making a major version 2.2?

**Reply by mtangoo** [reply]
What advantage will that offer over organic move to Yii3 from Yii2?

**Reply by lucaswitch** [reply]
It could be a flavor of Yii like we have the Basic App, the Advanced App and know you could create something like the Yii2 with some Yii3 packages.
I think the main idea is to keep yii2 packages updated alongside Yii3.

**Reply by mtangoo** [reply]
I think you missed the whole point. It is replacing the underlying packages in the framework with little to no change to user codebase. That is, instead of using Yii2's framework db, Yii will be using yii-db under the hood. Advantage is little maintenance costs as we will not be maintaining 2 dbs but one for Yii3, shared with Yii2

**Reply by fusivel** [reply]
@mtangoo The advantage would be of not touching in yii2 code and all not yet ready extensions, while not moving and rewriting to yii3 completely and, give the option to use those new packages instead, without upgrading to a "in-between" major version that will break some things. An idea, just that. Also maybe it will give less work to do... Certainly I am missing something that invalidate this. You tell.

**Reply by mtangoo** [reply]
>The advantage would be of not touching in yii2 code

Yii 2.2 is all about "touching" Yii2

> give the option to use those new packages instead, without upgrading to a "in-between" major version that will break some things

Well, direct upgrade to Yii3 for a huge codebase is not that easy.  Yii3 is a huge leap from concept of a framework in Yii2, which was monolithic. So "in-between" version will easy the burden and smoothen the process of transitioning.

Note that there will be Yii2.2 for people who a ready to break those little things and Yii2 for the those reluctant to break anything.

Not sure if there is anything there that you missed :)

**Reply by lucaswitch** [reply]
@mtangoo Got it, will it have some kind of configuration compatibility for example?

**Today** we mostly use the **yii\web\Request** to process incoming requests.
Does **yii\web\Request** will be rewritten or will be replaced by **Yii3 Request** package alongside some kind of wrapper, so in that way we can take the desired option on the main Yii configuration file and keep the Yii3 packages enabled by default.

**Reply by razvanphp** [reply]
@samdark should we bring (parts of) your [lib](https://github.com/samdark/yii2-psr-log-target), PSR-3 compatible logging to Yii 2.2 or is using yii3-log component easier to integrate?

**Comment by jiaweipan** [comment]
In our old big application, we did not use strict mode `declare(strict_types=1)`.
If we use the new Yii3.0 package, Does it mean that we need to refactor our application to comply with it?

**Reply by samdark** [reply]
> Strict typing applies to function calls made from within the file with strict typing enabled, not to the functions declared within that file. If a file without strict typing enabled makes a call to a function that was defined in a file with strict typing, the caller's preference (coercive typing) will be respected, and the value will be coerced.

https://www.php.net/manual/en/language.types.declarations.php#language.types.declarations.strict

**Comment by xepozz** [comment]
Also we may try to integrate yii-debug and yii-dev-panel. All of we need is to write collectors and add debug bootstrap.

**Comment by tomaszkane** [comment]
Replace routing will be super-cool, but it's complicated I suppose.

**Reply by samdark** [reply]
That would be a huge BC-break since Yii3 routing is different.

**Reply by tomaszkane** [reply]
Yes, and it's much better :-)

**Reply by mtangoo** [reply]
At that point one can move to Yii3 :)

**Reply by razvanphp** [reply]
Can we at least adopt PSR 7 or 15 so that roadrunner or frankenphp can work in worker mode?

Related: https://github.com/charlesportwoodii/yii2-psr7-bridge & #11328

**Comment by mtangoo** [comment]
>The hard decision here is to align PHP versions so that will require Yii 2.2 to be PHP 8.0+ at least.

No, it becomes even easier to decide. Whatever is Yii3 minimum, becomes Yii2.2 minimum. No?

**Reply by toMeloos** [reply]
I don't consider 8.0+ as a minimum requirement a bad thing given that it's also the minimum supported version of the language that everyone should be on: [https://www.php.net/supported-versions.php](https://www.php.net/supported-versions.php)

**Comment by mtangoo** [comment]
>https://github.com/yiisoft/db — to replace database-related functionality (could be tricky and raises questions about AR which isn't yet ready).

@terabytesoftw is there an ETA for AR?

**Reply by terabytesoftw** [reply]
AR as it is works very well, I have tested it 100%, @darkdef and I would like to improve the code even more, but that will take time to think and do.

**Reply by mtangoo** [reply]
@samdark so it should not be the problem!

**Reply by samdark** [reply]
Well... I can't say so. AR is complicated, refactoring is wanted and there's no release tagged yet.

**Reply by PowerGamer1** [reply]
Yii3 AR has severe inherent issues copy-pasted from Yii2: #19788.

**Reply by samdark** [reply]
Yes. It works but basically the same way with some fixes/refactoring.

**Reply by PowerGamer1** [reply]
> Yes. It works but basically the same way with some fixes/refactoring.

Yes, the same BAD way. @samdark Please come to #19788 - I am sure you don't want to inherit something like that in Yii3 (resolving it in Yii2 will automatically resolve it in Yii3).

**Comment by toMeloos** [comment]
What's the migration strategy? Can this be done without affecting applications built on top of Yii2? Can new Yii2 versions be released with these components swapped out under the hood without needing changesto our application code?

**Reply by terabytesoftw** [reply]
What do you think of opening a category on migrations from `Yii2` to `Yii3`, i could make an example of how to migrate an app, at least the steps, the db and ar api do not vary much in this regard.

**Reply by schmunk42** [reply]
@terabytesoftw That would be nice.

Also some Apps which come to my mind to test these changes with:
- Craft CMS
- Luya
- phd5
- HumHub

**Reply by terabytesoftw** [reply]
If precisely thinking about the great `Yii2` apps, we kept the `db` and `ar` `api` almost intact, and i like your suggestion, let's try one of them, it's a great idea.

**Reply by schmunk42** [reply]
I can do some testing with phd5. 

Just let me know when you think it's ready for testing.

**Reply by terabytesoftw** [reply]
You can try the 2.2 branch, it has cleaned almost all the deprecated methods, and i added `className()` again, so it should work without BC with the minimum version of `PHP 8.1` @schmunk42

**Comment by mtangoo** [comment]
Can we list Potential Yii3 packages that cn be used to replace current Yii2?
I believe with that list at least someone can pick one and tinker around

**Comment by mtangoo** [comment]
I missed this list by @samdark 

https://github.com/yiisoft/cache — to replace caching.
https://github.com/yiisoft/db — to replace database-related functionality (could be tricky and raises questions about AR which isn't yet ready).
https://github.com/yiisoft/rbac — to replace RBAC.
https://github.com/yiisoft/arrays — to replace array helper (partially).
https://github.com/yiisoft/log — to replace logger.
https://github.com/yiisoft/strings — to replace string helper (partially).
https://github.com/yiisoft/mutex — to replace mutex.
https://github.com/yiisoft/files — to replace file helper (partially).
https://github.com/yiisoft/json — to replace JSON helper.
https://github.com/yiisoft/network-utilities — to replace implementations of network-related stuff.
https://github.com/yiisoft/http — to replace status codes.
https://github.com/yiisoft/mailer — to replace mailer abstraction and drivers.
https://github.com/yiisoft/translator — to replace underlying implementation of Yii::t().
https://github.com/yiisoft/var-dumper — to replace the corresponding helper.
